File: mini-program/src/api/order.ts
// 订单/地址/上传相关接口
import { BASE_URL, http } from './request'

// ---------- 类型 ----------

export interface OrderListItem {
  id: number
  order_no: string
  cat_name: string
  status: string
  status_label: string
  total_price: number | null
  deposit_due: number
  making_due: number
  final_due: number
  paid_deposit: number
  paid_making: number
  paid_final: number
  cover_image_url: string
  queue_no: number
  created_at: string
}

export interface OrderDetail extends OrderListItem {
  images: string[]
  requirement: string
  address: { receiver?: string; phone?: string; region?: string; detail?: string }
  making_photos: string[]
  shipping_company: string
  tracking_no: string
  price_note: string
  refund_reason: string
  refund_amount: number
  cancel_reason: string
  coupon_amount: number
}

export interface AddressItem {
  id: number
  receiver: string
  phone: string
  region: string
  detail: string
  is_default: boolean
  created_at: string
}

export interface WorkItem {
  cat_name: string
  cover_image_url: string
  completed_at: string | null
}

// ---------- 订单 ----------

export function createOrder(payload: {
  cat_name: string
  images: string[]
  address_id: number
  requirement?: string
  coupon_id?: number
  referrer_user_id?: number
}) {
  return http.post<OrderListItem>('/orders', payload)
}

export function getMyOrders() {
  return http.get<OrderListItem[]>('/orders')
}

export function getOrderDetail(id: number) {
  return http.get<OrderDetail>(`/orders/${id}`)
}

export function cancelOrder(id: number, reason = '') {
  return http.post<OrderListItem>(`/orders/${id}/cancel?reason=${encodeURIComponent(reason)}`, {})
}

export function requestOrderRefund(id: number, reason = '') {
  return http.post<OrderListItem>(`/orders/${id}/refund-request?reason=${encodeURIComponent(reason)}`, {})
}

export function confirmReceive(id: number) {
  return http.post<OrderListItem>(`/orders/${id}/confirm-receive`, {})
}

// ---------- 地址簿 ----------

export function getAddresses() {
  return http.get<AddressItem[]>('/addresses')
}

export function createAddress(payload: Omit<AddressItem, 'id' | 'created_at'>) {
  return http.post<AddressItem>('/addresses', payload)
}

export function updateAddress(id: number, payload: Omit<AddressItem, 'id' | 'created_at'>) {
  return http.put<AddressItem>(`/addresses/${id}`, payload)
}

export function deleteAddress(id: number) {
  return http.post<{ detail: string }>(`/addresses/${id}/delete`, {})
}

// 便捷删除(DELETE 语义)
export function removeAddress(id: number) {
  return http.delete<{ detail: string }>(`/addresses/${id}`)
}

// ---------- 公开内容 ----------

export function getWorks() {
  return http.get<WorkItem[]>('/works', false)
}

export function getHomeSettings() {
  return http.get<{ promo_image_url: string }>('/home-settings', false)
}

// ---------- 用户图片上传 ----------

export function uploadUserImage(filePath: string): Promise<{ url: string }> {
  return new Promise((resolve, reject) => {
    uni.uploadFile({
      url: BASE_URL + '/user/uploads/image',
      filePath,
      name: 'file',
      header: { Authorization: `Bearer ${uni.getStorageSync('access_token') || ''}` },
      success: (res) => {
        if (res.statusCode >= 200 && res.statusCode < 300) {
          try {
            resolve(JSON.parse(res.data) as { url: string })
          } catch {
            reject(new Error('响应解析失败'))
          }
        } else {
          let detail = '上传失败'
          try {
            detail = (JSON.parse(res.data) as { detail?: string }).detail || detail
          } catch {
            /* ignore */
          }
          uni.showToast({ title: detail, icon: 'none' })
          reject(new Error(detail))
        }
      },
      fail: (err) => {
        uni.showToast({ title: '网络异常,请稍后重试', icon: 'none' })
        reject(new Error(err.errMsg))
      },
    })
  })
}

File: server/app/schemas/order.py
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field


# ---------- 用户端:创建订单 ----------
class OrderCreateIn(BaseModel):
    cat_name: str = Field(min_length=1, max_length=64)
    images: list[str] = Field(min_length=1, max_length=12)
    address_id: int
    requirement: str = Field(default="", max_length=2000)
    coupon_id: int | None = None
    referrer_user_id: int | None = None


class OrderListItem(BaseModel):
    id: int
    order_no: str
    cat_name: str
    status: str
    status_label: str = ""
    total_price: int | None = None
    deposit_due: int = 0
    making_due: int = 0
    final_due: int = 0
    paid_deposit: int = 0
    paid_making: int = 0
    paid_final: int = 0
    cover_image_url: str = ""
    queue_no: int = 0
    created_at: datetime


class OrderDetailOut(OrderListItem):
    images: list[str] = []
    requirement: str = ""
    address: dict = {}
    making_photos: list[str] = []
    shipping_company: str = ""
    tracking_no: str = ""
    price_note: str = ""
    refund_reason: str = ""
    refund_amount: int = 0
    cancel_reason: str = ""
    coupon_amount: int = 0


# ---------- 地址簿 ----------
class AddressIn(BaseModel):
    receiver: str = Field(min_length=1, max_length=64)
    phone: str = Field(min_length=5, max_length=20)
    region: str = Field(default="", max_length=128)
    detail: str = Field(default="", max_length=500)
    is_default: bool = False


class AddressOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    receiver: str
    phone: str
    region: str
    detail: str
    is_default: bool
    created_at: datetime


# ---------- 公开:作品画廊 / 首页配置 ----------
class WorkOut(BaseModel):
    cat_name: str
    cover_image_url: str
    completed_at: datetime | None = None


class HomeSettingsOut(BaseModel):
    promo_image_url: str = ""


class HomeSettingsIn(BaseModel):
    promo_image_url: str = Field(default="", max_length=500)


# ---------- 管理端 ----------
class PriceSetIn(BaseModel):
    total_price: int = Field(gt=0)
    note: str = Field(default="", max_length=255)


class ConfirmPaymentIn(BaseModel):
    stage: str = Field(pattern="^(deposit|making|final)$")
    # 允许 0:低总价订单的制作定金可为 0(免付阶段,确认即推进状态)
    amount: int = Field(ge=0)
    note: str = Field(default="", max_length=255)


class MakingPhotosIn(BaseModel):
    photos: list[str] = Field(max_length=3)


class SetCoverIn(BaseModel):
    image_url: str = Field(min_length=1)


class ShipIn(BaseModel):
    shipping_company: str = Field(min_length=1, max_length=64)
    tracking_no: str = Field(min_length=1, max_length=64)


class RefundIn(BaseModel):
    amount: int = Field(gt=0)
    reason: str = Field(default="", max_length=255)


class CancelIn(BaseModel):
    reason: str = Field(default="", max_length=255)


class AdminOrderListItem(BaseModel):
    id: int
    order_no: str
    status: str
    status_label: str = ""
    cat_name: str
    user_nickname: str = ""
    user_phone: str | None = None
    total_price: int | None = None
    deposit_due: int = 0
    making_due: int = 0
    final_due: int = 0
    paid_deposit: int = 0
    paid_making: int = 0
    paid_final: int = 0
    address: dict = {}
    images: list[str] = []
    cover_image_url: str = ""
    referrer_user_id: int | None = None
    queue_no: int = 0
    created_at: datetime


class AdminOrderDetailOut(AdminOrderListItem):
    requirement: str = ""
    making_photos: list[str] = []
    shipping_company: str = ""
    tracking_no: str = ""
    price_note: str = ""
    refund_reason: str = ""
    refund_amount: int = 0
    cancel_reason: str = ""
    coupon_id: int | None = None
    coupon_amount: int = 0